bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/test/java/com/google/cloud/bigtable/mirroring/hbase1_x/TestMirroringTable.java

```diff
@@ -0,0 +1,1411 @@
+
+    // successful secondary reads were reported
+    verify(mismatchDetector, times(1))
+        .batch(Arrays.asList(get2), new Result[] {get2Result}, new Result[] {get2Result});
+
+    // no read errors reported
+    verify(mismatchDetector, never())
+        .batch(ArgumentMatchers.<Get>anyList(), any(IOException.class));
+  }
+
+  @Test
+  public void testConditionalWriteHappensWhenConditionIsMet() throws IOException {
+    Put put = new Put("r1".getBytes());
+    when(primaryTable.checkAndMutate(
+            any(byte[].class),
+            any(byte[].class),
+            any(byte[].class),
+            eq(CompareOp.EQUAL),
+            any(byte[].class),
+            any(RowMutations.class)))
+        .thenReturn(true);
+
+    mirroringTable.checkAndPut(
+        "r1".getBytes(), "f1".getBytes(), "q1".getBytes(), "v1".getBytes(), put);
+    executorServiceRule.waitForExecutor();
+
+    ArgumentCaptor<RowMutations> argument = ArgumentCaptor.forClass(RowMutations.class);
+    verify(secondaryTable, times(1)).mutateRow(argument.capture());
+    assertThat(argument.getValue().getRow()).isEqualTo("r1".getBytes());
+    assertThat(argument.getValue().getMutations().get(0)).isEqualTo(put);
+  }
+
+  @Test
+  public void testConditionalWriteDoesntHappenWhenConditionIsNotMet() throws IOException {
+    Put put = new Put("r1".getBytes());
+    when(primaryTable.checkAndMutate(
+            any(byte[].class),
+            any(byte[].class),
+            any(byte[].class),
+            eq(CompareOp.EQUAL),
+            any(byte[].class),
+            any(RowMutations.class)))
+        .thenReturn(false);
+
+    mirroringTable.checkAndPut(
+        "r1".getBytes(), "f1".getBytes(), "q1".getBytes(), "v1".getBytes(), put);
+    executorServiceRule.waitForExecutor();
+
+    verify(secondaryTable, never()).mutateRow(any(RowMutations.class));
+  }
+
+  @Test
+  public void testCheckAndPut() throws IOException {
+    Put put = new Put("r1".getBytes());
+    when(primaryTable.checkAndMutate(
+            any(byte[].class),
+            any(byte[].class),
+            any(byte[].class),
+            any(CompareOp.class),
+            any(byte[].class),
+            any(RowMutations.class)))
+        .thenReturn(true);
+
+    mirroringTable.checkAndPut(
+        "r1".getBytes(),
+        "f1".getBytes(),
+        "q1".getBytes(),
+        CompareOp.GREATER_OR_EQUAL,
+        "v1".getBytes(),
+        put);
+    executorServiceRule.waitForExecutor();
+
+    verify(secondaryTable, times(1)).mutateRow(any(RowMutations.class));
+  }
+
+  @Test
+  public void testCheckAndDelete() throws IOException {
+    Delete delete = new Delete("r1".getBytes());
+    when(primaryTable.checkAndMutate(
+            any(byte[].class),
+            any(byte[].class),
+            any(byte[].class),
+            any(CompareOp.class),
+            any(byte[].class),
+            any(RowMutations.class)))
+        .thenReturn(true);
+
```

Comment:
Explain what are the 2 checkAndDeletes doing and what do we expect at the end?

Reply:
We've removed one of the checks.